### ISSUE TITLE
"pure virtual method called" on Adreno 330

### ISSUE THREAD

**Issue by sivagnanamn** [CNugteren/CLBlast#178]
While doing "make alltuners" on Adreno 330, it stopped after CopyPadMatrix with the error below: 

```
[ RESULT   ] CopyPadMatrix;     18.6 ms;PAD_DIMX 32;PAD_DIMY 8;PAD_WPTX 4;PAD_WPTY 2;PRECISION 3232;
[ RESULT   ] CopyPadMatrix;     30.5 ms;PAD_DIMX 32;PAD_DIMY 8;PAD_WPTX 4;PAD_WPTY 4;PRECISION 3232;

[----------] Printing best result to stdout
[     BEST ] CopyPadMatrix;      4.1 ms;PAD_DIMX 32;PAD_DIMY 8;PAD_WPTX 1;PAD_WPTY 1;PRECISION 3232;

[----------] Printing best result in database format to stdout
{ "QUALCOMM Adreno(TM)", { {"PAD_DIMX",32}, {"PAD_DIMY",8}, {"PAD_WPTX",1}, {"PAD_WPTY",1}, {"PRECISION",3232} } }
[ -------> ] 4.14 ms or 4.0 GB/s

[----------] Printing results to file in JSON format

[==========] End of the tuning process

pure virtual method called
terminate called without an active exception
make[3]: *** [CMakeFiles/alltuners] Aborted
make[2]: *** [CMakeFiles/alltuners.dir/all] Error 2
make[1]: *** [CMakeFiles/alltuners.dir/rule] Error 2
```
Without tuning, while using sgemm in CLBlast I'm getting CL_OUT_OF_HOST_MEMORY 

`CLBlast: OpenCL error: clBuildProgram: -6`

**Comment by sivagnanamn** [comment]
Is it possible to tune only the kernels that I wish to try out? ex: make sgemmtuner

**Comment by CNugteren** [comment]
The `make alltuners` is just a shorthand from CMake for calling all the possible tuners manually. It will thus just execute all tuners with all precision options, so you can just run them individually, e.g.:
```
clblast_tuner_copy_fast -precision 32
clblast_tuner_copy_fast -precision 64
clblast_tuner_copy_fast -precision 3232
clblast_tuner_copy_fast -precision 6464
clblast_tuner_copy_fast -precision 16
clblast_tuner_copy_pad -precision 32
clblast_tuner_copy_pad -precision 64
clblast_tuner_copy_pad -precision 3232
clblast_tuner_copy_pad -precision 6464
clblast_tuner_copy_pad -precision 16
clblast_tuner_transpose_fast -precision 32
clblast_tuner_transpose_fast -precision 64
(etc.)
```

The `CL_OUT_OF_HOST_MEMORY` error you are getting is probably just a genuine out of memory error I guess. Can you please double check this and reduce the input sizes? The GEMM kernel might also allocate temporary buffers of equal size, so keep that in mind.

The problem with `pure virtual method called` I'm not sure of. Please run the tuners individually and check which one(s) fail and when, and then optionally re-run them under a debugger (e.g. gdb).

**Comment by sivagnanamn** [comment]
These are the sizes of different matrices that I tired using sgemm in CLBlast. 
```
M: 16 N: 200704 K: 27 
M: 32 N: 50176 K: 144 
M: 64 N: 12544 K: 288 
M: 128 N: 3136 K: 576 
M: 256 N: 784 K: 1152 
M: 512 N: 196 K: 2304 
M: 1024 N: 49 K: 4608 
M: 256 N: 49 K: 9216 
M: 1 N: 1029 K: 12544 
```
I have 300 MB of free GPU memory & 1GB of RAM free during execution always. But still I get  CL_OUT_OF_HOST_MEMORY for all the matrices

> CLBlast: OpenCL error: clBuildProgram: -6

**Comment by CNugteren** [comment]
The message `CL_OUT_OF_HOST_MEMORY` suggests host (CPU) memory problems. The message `clBuildProgram` suggests problems with compiling the kernel. So perhaps the kernel is too complicated to be compiled by the Adreno compiler with 1GB of RAM?

Could you maybe compile CLBlast with the clients and the tests (-DCLIENTS=ON -DTESTS=ON) and then run all or individual tests? If only the routines with complicated kernels fail (i.e. the GEMM family), then that might be valid indeed. The tuner would then be able to find cases for which the OpenCL compiler doesn't fail. Of course this is just a a hypothesis, to be tested by you :-)

**Comment by sivagnanamn** [comment]
@CNugteren Thanks again for your suggestions. Tried testing gemm alone after make with -DTESTS=ON. All tests failed with CLBlast: OpenCL error: clBuildProgram: -6 ( CL_OUT_OF_HOST_MEMORY )
```
* Completed all test-cases for this routine. Results:
   0 test(s) passed
   0 test(s) skipped
   512 test(s) failed
```
Other tests like axpy & dot operations passed. 

While running GEMM TUNER, few routines fail and most of them pass.

```
[ RUN      ] Running Xgemm
[       OK ] Completed Xgemm (1770.2 ms) - 52 out of 360
[   FAILED ] Kernel Xgemm failed
[   FAILED ]   catched exception: Internal OpenCL error: -6
[   FAILED ] Xgemm;      0.0 ms;   MWG 16;   NWG 32;   KWG 32; MDIMC 16; NDIMC 32; MDIMA 16; NDIMB 32;    KWI 2;    VWM 1;    VWN 1;   STRM 0;   STRN 0;     SA 0;     SB 0;PRECISION 32;
[ RUN      ] Running Xgemm
[   FAILED ] Kernel Xgemm failed
[   FAILED ]   catched exception: Internal OpenCL error: -54
[  WARNING ] Results differ: L2 norm is 2.14e+08
[   FAILED ] Xgemm;      0.0 ms;   MWG 16;   NWG 32;   KWG 32; MDIMC 16; NDIMC 32; MDIMA 16; NDIMB 32;    KWI 2;    VWM 1;    VWN 1;   STRM 0;   STRN 0;     SA 1;     SB 1;PRECISION 32;
[ RUN      ] Running Xgemm
[       OK ] Completed Xgemm (434.7 ms) - 55 out of 360
[ RUN      ] Running Xgemm
[       OK ] Completed Xgemm (437.1 ms) - 56 out of 360
```
Could you please suggest any other OpenCL gemm implementation that can work on Adreno 330 with OpenCL 1.1, 2GB RAM & ~700 MB GPU memory? Cannot install clBlas since it requires minimum OpenCL 1.2.

**Comment by CNugteren** [comment]
Let's try to make it work with CLBlast :-) Can you run the full GEMM tuner and preferably also the others and report back to me the JSON files they produce? I'll update the library and hopefully the issue is solved.

**Comment by sivagnanamn** [comment]
@CNugteren You're a rockstar.
I've attached the GEMM Tuner .json files. 
I'm getting "pure virtual method called" error when running alltuners. Will update the others ASAP.

[GEMM_Tuner.zip](https://github.com/CNugteren/CLBlast/files/1182170/GEMM_Tuner.zip)

**Comment by sivagnanamn** [comment]
Here are the other tuner output json files. Contains the outputs of below tuners

```
clblast_tuner_copy_fast
clblast_tuner_copy_pad
clblast_tuner_transpose_fast
clblast_tuner_transpose_pad
clblast_tuner_xaxpy
clblast_tuner_xdot
clblast_tuner_xgemm_direct
clblast_tuner_xgemv
clblast_tuner_xger
[Tuner_json_files.zip](https://github.com/CNugteren/CLBlast/files/1182216/Tuner_json_files.zip)

```

**Comment by CNugteren** [comment]
Thanks, I will add those results soon. In the meantime, you can try yourself by running the following from the CLBlast root:
```python scripts/database/database.py /path/to/json/files .``` (second argument is the path to CLBlast root)
Afterwards you can recompile the library and try the GEMM test again.

**Comment by sivagnanamn** [comment]
Thanks again, did that and recompiled the library. Now I'm not getting any error. It takes approx 800ms for M: 16 N: 200704 K: 27  matrix. Attached the latest GEMM json files.
[GEMM_Tuner_updated.zip](https://github.com/CNugteren/CLBlast/files/1182597/GEMM_Tuner_updated.zip)

**Comment by CNugteren** [comment]
I've added your tuning results to the main repository, it should at least now work out of the box on Adreno 330 GPUs.

I'm a bit worried about the performance numbers you report though. You are using very rectangular matrices, which the library is not optimised for (but it is on the agenda #169). From my calculations you only get ((2 * 16 * 200704 * 27) / 0.8) * 1e-9 = 0.2 GFLOPS, not that much. Could you enable the included clients (`-DCLIENTS=ON`, see README) and run `clblast_client_xgemm` with some more common sizes (e.g. 512x512x512) just to see that performance is at least OK for more regular cases.

**Comment by sivagnanamn** [comment]
Thanks for your time. Appreciate it.

Below is my result for

**Comment by sivagnanamn** [comment]
Thanks for your time. Appreciate it.

Unfortunately the data I'm dealing with are im2col outputs, so its mostly rectangular matrices.

Below is the result of  ```clblast_client_xgemm``` for regular size (512x512x512) 

```
* Options given/available:
    -m 512 [=default]
    -n 512 [=default]
    -k 512 [=default]
    -layout 101 (row-major) [=default]
    -transA 111 (regular) [=default]
    -transB 111 (regular) [=default]
    -lda 512 [=default]
    -ldb 512 [=default]
    -ldc 512 [=default]
    -offa 0 [=default]
    -offb 0 [=default]
    -offc 0 [=default]
    -alpha 2.00 [=default]
    -beta 2.00 [=default]
    -platform 0 [=default]
    -device 0 [=default]
    -precision 32 (single) [=default]
    -cblas 1 [=default]
    -step 1 [=default]
    -num_steps 0 [=default]
    -runs 10 [=default]
    -h [false]
    -q [false]
    -no_abbrv [false]
    -warm_up [false]

                                                                                                                                             | <--       CLBlast       --> | <--      CPU BLAS       --> |
        m;        n;        k;   layout;   transA;   transB;      lda;      ldb;      ldc;     offa;     offb;     offc;    alpha;     beta;     ms_1; GFLOPS_1;    GBs_1;     ms_3; GFLOPS_3;    GBs_3
      512;      512;      512;      101;      111;      111;      512;      512;      512;        0;        0;        0;     2.00;     2.00;    34.69;      7.7;      0.1;   253.49;      1.1;      0.0
```
Wiht the updated library, I'm getting `CLBlast: OpenCL error: clEnqueueNDRangeKernel: -54` (`CL_INVALID_WORK_GROUP_SIZE`) for `M: 256 N: 49 K: 9216`.

**Comment by CNugteren** [comment]
Sorry about that, but that should be fixed now thanks to #179.

**Comment by sivagnanamn** [comment]
Cool, its working now.
